stdlib/internal/SwiftSyntax/Syntax.swift

```diff
@@ -0,0 +1,183 @@
+//===-------------------- Syntax.swift - Syntax Protocol ------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// A Syntax node represents a tree of nodes with tokens at the leaves.
+/// Each node has accessors for its known children, and allows efficient
+/// iteration over the children through its `children` property.
+public class Syntax: CustomStringConvertible {
+  /// The root of the tree this node is currently in.
+  internal let _root: SyntaxData
+  
+  /// The data backing this node.
+  /// - note: This is unowned, because the reference to the root data keeps it
+  ///         alive. This means there is an implicit relationship -- the data
+  ///         property must be a descendent of the root. This relationship must
+  ///         be preserved in all circumstances where Syntax nodes are created.
+  internal unowned var data: SyntaxData
+
+#if DEBUG
+  func validate() {
```

Comment:
What's the outcome of the validation? Should it throw or return error if validation fails?